pycbc/detector/space.py

```diff
@@ -0,0 +1,899 @@
+        orbit_start_idx = numpy.argwhere(hp.sample_times.numpy() >= orbit_start_time)[0][0]
+        hp = hp[orbit_start_idx:]
+        hc = hc[orbit_start_idx:]
+
+    return hp, hc
+
+def postprocess(tdi_dict, remove_garbage=False, t0=1e4):
+    """
+    Apply start times to TDI channels and cut if needed.
+
+    Parameters
+    ----------
+    tdi_dict : dict
+        The TDI channels, formatted as a dictionary of TimeSeries arrays
+        keyed by the channel label.
+
+    remove_garbage : bool, str (optional)
+        Flag whether to remove data from the edges of the channels. If True,
+        time length t0 is cut from the start and end. If 'zero', time length
+        t0 is zeroed at the start and end. If False, channels are unmodified.
+        Default False.
+
+    t0 : float (optional)
+        Time in seconds to cut/zero from data if remove_garbage is True/'zero'.
+        Default 1e4.
+    """
+    for chan in tdi_dict.keys():
+        if remove_garbage:
+            dt = tdi_dict[chan].delta_t
+            pad_idx = int(t0/dt)
+            if remove_garbage == 'zero':
+                # zero the edge data
+                tdi_dict[chan][:pad_idx] = 0
+                tdi_dict[chan][-pad_idx:] = 0
+            elif type(remove_garbage) == bool:
+                # cut the edge data
+                slc = slice(pad_idx, -pad_idx)
+                tdi_dict[chan] = tdi_dict[chan][slc]
+            else:
+                raise ValueError('remove_garbage arg must be a bool ' +
+                                 'or "zero"')
+
+    return tdi_dict
+
+class AbsSpaceDet(ABC):
+    """
+    Abstract base class to set structure for space detector classes.
+
+    Parameters
+    ----------
+    reference_time : float (optional)
+        The reference time in seconds of the signal in the SSB frame. This
+        will correspond to the zero point of the input signal. By default,
+        a reference time is automatically assigned such that the input signal
+        starts at zero. Default None.
+
+    apply_offset : bool (optional)
+        Flag whether to shift the times of the input waveforms by
+        a given value. Some backends require this such that the
+        detector is oriented correctly at t = 0. Default False.
+
+    offset : float (optional)
+        The time in seconds by which to offset the input waveform if
+        apply_offset is True. Default 7365189.431698299.
+    """
+    def __init__(self, apply_offset=False, 
+                 offset=TIME_OFFSET_20_DEGREES, **kwargs):
+        # specify whether to apply offsets to GPS times
+        if apply_offset:
+            self.offset = offset
+        else:
+            self.offset = 0.
+
+    @property
+    @abstractmethod
+    def sky_coords(self):
+        """
+        List the sky coordinate names for the detector class.
+        Raises a NotImplementedError if not specified in child class.
+        """
+        return
+
+    @abstractmethod
+    def orbits_init(self):
```

Comment:
I'm not sure if `get_links` or `orbits_init` should be in the base class or not. What does get_links return? For oribts_init, why wouldn't that just be a part of the init of the class itself. It doesn't seem to take any arguments. 

The standard arguments and expected return should be specified for the base class methods. Yes, it's not enforced, but it's in part so that people (and future developers) know what the API is supposed to be. Project wave for example should specify the minimum interface that can be expected (but not for example any optional argument or extensions).

Reply:
`get_links` in the child classes returns the GW projected to the six laser links. This would be the intermediary step before calculating the TDI combinations. Initially, I thought it would be useful to have this be common to all classes in case, e.g., someone wanted to homebrew their own TDI combinations. I can see in retrospect, though, that this shouldn't be required. The same is true for `orbits_init`, although I've left them as methods for now just to keep things more organized.